torch/csrc/autograd/profiler_kineto.cpp

```diff
@@ -0,0 +1,298 @@
+#include <torch/csrc/autograd/profiler_kineto.h>
+
+#include <torch/csrc/jit/frontend/tracer.h>
+#include <torch/csrc/jit/runtime/operator.h>
+
+#include <sstream>
+
+#ifndef _WIN32
+#include <pthread.h>
+#endif
+
+#ifdef USE_KINETO
+#include "libkineto.h"
+#endif
+
+namespace torch { namespace autograd { namespace profiler {
+
+#ifdef USE_KINETO
+namespace {
+// TODO: consider TLS
+uint64_t next_correlation_id() {
+  static std::atomic<uint64_t> corr_id_ {1};
+  return corr_id_++;
+}
+
+std::string shapesToStr(const std::vector<std::vector<int64_t>>& shapes);
+
+struct TORCH_API KinetoThreadLocalState : public ProfilerThreadLocalState {
+  using ProfilerThreadLocalState::ProfilerThreadLocalState;
+  virtual ~KinetoThreadLocalState() override = default;
+
+  virtual void reportClientActivity(
+      const at::RecordFunction& fn,
+      const at::ObserverContext* observer_ctx) override {
+    auto ctx = dynamic_cast<const KinetoObserverContext*>(observer_ctx);
```

Comment:
I think it's safe to put `static_cast` here and everywhere because we rely on the right function being called with the right arg. This kind of argument co-variance is common in callbacks.

Reply:
will upd